File: src/attributes.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { matchClock, initial, step, resetPitch } from "./game.js";
import { pickup, damage, featureStep } from "./features.js";
import {
  movementSpeed,
  velocityUnit,
  tackleThreshold,
  hitDamage,
  randomByte,
} from "./attributes.js";

test("Amiga tackle thresholds cover every facing, keeper bonus and jumping", () => {
  const s = initial(),
    p = s.players[7],
    q = s.players[16];
  p.action = 1;
  for (let d = 0; d < 8; d++) {
    q.fx = Math.cos((d * Math.PI) / 4);
    q.fz = Math.sin((d * Math.PI) / 4);
    assert.equal(
      tackleThreshold(p, q, false),
      [152, 148, 144, 140, 136, 140, 144, 148][d],
    );
  }
  q.fx = -1;
  q.fz = 0;
  assert.equal(tackleThreshold(p, q, true), 123);
  q.action = 2;
  assert.equal(tackleThreshold(p, q, true), 139);
});

test("reference RNG matches word-register trace including carries", () => {
  const s = initial();
  for (const expected of [
    2224703967, 3051111946, 980885458, 4032023480, 717936405, 454968731,
    1172912480, 1627913718,
  ]) {
    assert.equal(randomByte(s), expected & 255);
    assert.equal(s.rng[0], expected);
  }
});

test("failed tackle consumes its contact without a retry on later frames", () => {
  const s = initial();
  s.players.forEach((p) => (p.stun = 10));
  Object.assign(s.players[7], { x: 0, z: 0, stun: 0, fx: 1, fz: 0 });
  Object.assign(s.players[16], {
    x: 0.7,
    z: 0,
    stun: 0,
    cooldown: 10,
    fx: -1,
    fz: 0,
  });
  Object.assign(s.ball, { owner: 16, x: 0.7, z: 0, h: 1 });
  step(s, 1 / 60, { tackle: true }, [true, true]); // First reference random byte is 223, threshold 136.
  assert.equal(s.ball.owner, 16);
  assert.equal(s.players[16].health, 100);
  const rng = [...s.rng];
  for (let n = 0; n < 4; n++) {
    Object.assign(s.players[16], {
      x: s.players[7].x + 0.3,
      z: s.players[7].z,
    });
    step(s, 1 / 60, {}, [true, true]);
  }
  assert.deepEqual(s.rng, rng);
  assert.equal(s.players[16].health, 100);
});

test("damage uses power and stamina, deteriorates all attributes, and respects the floor", () => {
  const s = initial(),
    p = s.players[7],
    q = s.players[16];
  assert.equal(hitDamage(p, q), 9);
  p.stats[5] = 250;
  q.stats.fill(200);
  assert.equal(hitDamage(p, q), 12);
  damage(s, 7, 16);
  assert.equal(q.health, 100 - (12 * 100) / 128);
  assert.deepEqual(q.stats, Array(8).fill(194));
  q.stun = 0;
  q.stats.fill(100);
  q.stats[6] = 250;
  p.stats[5] = 100;
  assert.equal(hitDamage(p, q), 1);
  damage(s, 7, 16);
  assert.equal(q.stats[0], 100);
});

test("power and armour attributes restore in reference order after a hit and expiry", () => {
  const s = initial(),
    p = s.players[7];
  p.stats.fill(180);
  pickup(s, 7, 4);
  assert.deepEqual(p.stats, Array(8).fill(250));
  pickup(s, 7, 15);
  damage(s, 16, 7);
  assert.equal(p.gear, 0);
  assert.equal(p.stats[1], 250);
  matchClock(s, 6);
  assert.deepEqual(p.stats, Array(8).fill(180));
  pickup(s, 7, 19);
  assert.equal(p.stats[5], 250);
  s.tick = 1;
  resetPitch(s);
  assert.equal(s.players[7].stats[5], 250);
  pickup(s, 7, 11);
  assert.deepEqual(s.players[7].stats, Array(8).fill(100));
  assert.equal(s.players[7].gear, 0);
});

test("movement uses strict original speed boundaries for running, carrying, sliding and jumping", () => {
  const p = initial().players[7];
  for (const [stat, run, slide, jump] of [
    [100, 4, 5, 4],
    [140, 4, 5, 4],
    [141, 5, 6, 5],
    [170, 5, 6, 5],
    [171, 5, 7, 6],
    [200, 5, 7, 6],
    [201, 6, 8, 7],
    [250, 6, 8, 7],
  ]) {
    p.stats[3] = stat;
    p.action = 0;
    assert.equal(movementSpeed(p, true), run * velocityUnit);
    assert.equal(movementSpeed(p, false), (run + 1) * velocityUnit);
    p.action = 1;
    assert.equal(movementSpeed(p, false), slide * velocityUnit);
    assert.equal(movementSpeed(p, false, true), 8 * velocityUnit);
    p.action = 2;
    assert.equal(movementSpeed(p, false), jump * velocityUnit);
  }
});
test("slow sets a fast player to the baseline speed and expiry restores it", () => {
  const s = initial(),
    p = s.players[7];
  p.stats[3] = 250;
  assert.equal(movementSpeed(p, false), 7 * velocityUnit);
  pickup(s, 16, 6);
  assert.equal(movementSpeed(p, false), 5 * velocityUnit);
  matchClock(s, 6);
  assert.equal(movementSpeed(p, false), 7 * velocityUnit);
});

test("match time and powers share second pulses; medical pauses only match time", () => {
  const s = initial();
  pickup(s, 7, 4);
  matchClock(s, 0.99);
  assert.equal(s.time, 90);
  assert.equal(s.effect.time, 6);
  matchClock(s, 0.01);
  assert.equal(s.time, 89);
  assert.equal(s.effect.time, 5);
  s.players[16].injury = 6;
  matchClock(s, 5);
  assert.equal(s.time, 89);
  assert.equal(s.effect.kind, 0);
  assert.deepEqual(s.players[7].stats, Array(8).fill(100));
  s.players[16].injury = 0;
  s.pause = 2;
  pickup(s, 7, 10);
  matchClock(s, 1);
  assert.equal(s.time, 89);
  assert.equal(s.effect.time, 5);
});

File: server/attributes_test.go
package main

import "testing"

func TestSelectionSwitchesDuringActionsAndBreaksTiesByRosterOrder(t *testing.T) {
	s := initial()
	for i := range s.Players {
		s.Players[i].Stun = 10
	}
	s.Ball.Owner = -1
	s.Ball.X, s.Ball.Z = 0, 0
	p := &s.Players[7]
	p.X, p.Z, p.Stun, p.Action, p.ActionTime = 2, 0, 0, 1, 0.3
	s.Players[0].X, s.Players[0].Z, s.Players[0].Stun = 2.5, 0, 0
	s.Players[8].X, s.Players[8].Z, s.Players[8].Stun = 1.5, 0, 0
	s.Controlled[0] = 7
	s.selectPlayers()
	if s.Controlled[0] != 8 {
		t.Fatal("action or goalkeeper bias prevented switching")
	}
	p.X = 1.5
	s.Controlled[0] = 7
	s.selectPlayers()
	if s.Controlled[0] != 8 {
		t.Fatal("equal distance should favor later roster index")
	}
	s.Ball.Owner = 7
	s.selectPlayers()
	if s.Controlled[0] != 7 {
		t.Fatal("ball owner must remain controlled")
	}
}

func TestAmigaTackleAndDamageRules(t *testing.T) {
	s := initial()
	p, q := &s.Players[7], &s.Players[16]
	p.Action = 1
	q.FX = -1
	if tackleThreshold(p, q, false) != 136 || tackleThreshold(p, q, true) != 123 {
		t.Fatal("front or goalkeeper threshold")
	}
	q.Action = 2
	q.jumping = true
	if tackleThreshold(p, q, true) != 139 {
		t.Fatal("jump penalty")
	}
	q.jumping = false
	if tackleThreshold(p, q, true) != 123 {
		t.Fatal("landing retains jump penalty")
	}
	q.Action = 0
	q.FX = 1
	if tackleThreshold(p, q, false) != 152 {
		t.Fatal("rear penalty")
	}
	if hitDamage(p, q) != 9 {
		t.Fatal("base damage")
	}
	p.Stats[5] = 250
	for i := range q.Stats {
		q.Stats[i] = 200
	}
	if hitDamage(p, q) != 12 {
		t.Fatal("power/stamina damage")
	}
	s.damage(7, 16)
	if q.Health != 100-12.*100/128 {
		t.Fatal("energy scaling")
	}
	for _, v := range q.Stats {
		if v != 194 {
			t.Fatal("attribute deterioration")
		}
	}
}
func TestAmigaRandomTrace(t *testing.T) {
	s := initial()
	for _, expected := range []uint32{2224703967, 3051111946, 980885458, 4032023480, 717936405, 454968731, 1172912480, 1627913718} {
		if s.randomByte() != int(expected&255) || s.RNG[0] != expected {
			t.Fatal("word-register trace", expected, s.RNG)
		}
	}
}
func TestFailedTackleDoesNotReroll(t *testing.T) {
	s := isolated()
	s.Players[16] = Player{Stats: defaultStats(), Health: 100, Team: 1, X: .7, FX: -1, Cooldown: 10}
	s.Ball = Ball{Owner: 16, X: .7, H: 1}
	s.step(dt, [2]Input{{Tackle: true, X: 1}, {}})
	s.step(dt, [2]Input{})
	if s.Ball.Owner != 16 || s.Players[16].Health != 100 {
		t.Fatal("failed roll took possession")
	}
	rng := s.RNG
	for n := 0; n < 4; n++ {
		s.Players[16].X = s.Players[7].X + .3
		s.Players[16].Z = s.Players[7].Z
		s.step(dt, [2]Input{})
	}
	if s.RNG != rng || s.Players[16].Health != 100 {
		t.Fatal("tackle retried")
	}
}
func TestPowerArmourRestoration(t *testing.T) {
	s := initial()
	p := &s.Players[7]
	for i := range p.Stats {
		p.Stats[i] = 180
	}
	s.pickup(7, 4)
	s.pickup(7, 15)
	s.damage(16, 7)
	if p.Gear != 0 || p.Stats[1] != 250 {
		t.Fatal("armour restoration under power")
	}
	s.matchClock(6)
	for _, v := range p.Stats {
		if v != 180 {
			t.Fatal("power restoration")
		}
	}
	s.pickup(7, 19)
	s.Tick = 1
	s.resetPitch()
	if s.Players[7].Stats[5] != 250 {
		t.Fatal("goal lost attributes")
	}
	s.pickup(7, 11)
	if s.Players[7].Stats != defaultStats() || s.Players[7].Gear != 0 {
		t.Fatal("full energy reset")
	}
}

func TestMovementTableBoundaries(t *testing.T) {
	s := initial()
	p := &s.Players[7]
	for _, c := range [][4]int{{100, 4, 5, 4}, {140, 4, 5, 4}, {141, 5, 6, 5}, {170, 5, 6, 5}, {171, 5, 7, 6}, {200, 5, 7, 6}, {201, 6, 8, 7}, {250, 6, 8, 7}} {
		p.Stats[3] = c[0]
		p.Action = 0
		if movementSpeed(p, true, false) != float64(c[1])*velocityUnit || movementSpeed(p, false, false) != float64(c[1]+1)*velocityUnit {
			t.Fatal("run/carry boundary", c)
		}
		p.Action = 1
		if movementSpeed(p, false, false) != float64(c[2])*velocityUnit || movementSpeed(p, false, true) != 8*velocityUnit {
			t.Fatal("slide boundary", c)
		}
		p.Action = 2
		if movementSpeed(p, false, false) != float64(c[3])*velocityUnit {
			t.Fatal("jump boundary", c)
		}
	}
	p.Action = 0
	p.Stats[3] = 250
	s.pickup(16, 6)
	if movementSpeed(p, false, false) != 5*velocityUnit {
		t.Fatal("slow baseline")
	}
	s.matchClock(6)
	if movementSpeed(p, false, false) != 7*velocityUnit {
		t.Fatal("slow restoration")
	}
}

func TestSharedMatchClock(t *testing.T) {
	s := initial()
	s.pickup(7, 4)
	s.matchClock(.99)
	if s.Time != 90 || s.Effect.Time != 6 {
		t.Fatal("early pulse")
	}
	s.matchClock(.01)
	if s.Time != 89 || s.Effect.Time != 5 {
		t.Fatal("shared pulse")
	}
	s.Players[16].Injury = 6
	s.matchClock(5)
	if s.Time != 89 || s.Effect.Kind != 0 || s.Players[7].Stats != defaultStats() {
		t.Fatal("medical pause power expiry")
	}
	s.Players[16].Injury = 0
	s.Pause = 2
	s.pickup(7, 10)
	s.matchClock(1)
	if s.Time != 89 || s.Effect.Time != 5 {
		t.Fatal("goal pause")
	}
}
